Improve handling of SSL metadata in ConnectionMetaData, proxied connections and forwarded requests #10775

gregw opened the pull request:
Fix #10775:
+ Added an Attributes.Synthetic mechanism to better (and mor commonly) handle additional attributes, whilst meeting the new TCK requirement that they can be deleted or replaced.
+ Made SslSessionData a reusable type provided by EndPoint
+ Added the EndPoint.Securable subtype that can provide a SslSessionData instance for secure EndPoints
+ Updated ForwardedRequestCustomizer, SecureRequestCustomizer to understand Securable and SslSessionData

gregw committed:
Make ConnectionMetaData isSecure work correctly

WIP

gregw committed:
Make ConnectionMetaData isSecure work correctly

WIP

olamy committed:
add missing license header and use modify source snapshot

Signed-off-by: Olivier Lamy <olamy@apache.org>

gregw committed:
Make ConnectionMetaData isSecure work correctly

gregw committed:
Make ConnectionMetaData isSecure work correctly

WIP

gregw committed:
Improve handling of SSL metadata in ConnectionMetaData, proxied connections and forwarded requests

+ Added an Attributes.Synthetic mechanism to better (and mor commonly) handle additional attributes, whilst meeting the new TCK requirement that they can be deleted or replaced.
+ Made SslSessionData a reusable type provided by EndPoint
+ Added the EndPoint.Securable subtype that can provide a SslSessionData instance for secure EndPoints
+ Updated ForwardedRequestCustomizer, SecureRequestCustomizer to understand Securable and SslSessionData

gregw commented:
@sbordet @joakime This PR became a bit of a monster.....  
I was just going to fix the handling of SSL isSecure for proxied requests, but then I realized that our handling of synthetic attributes (used for SSL metadata) was mostly wrong against the latest TCK (which requires that all attributes can be deleted or modified), plus there were some incorrect usages of "jakarta" in attribute names.    So I had to refactor that mechanism.

I have then pushed down the concept of `SslSessionData` to `EndPoint.Securable`, so we can ask any endpoint if it is secure and metadata about that.  Now SslEndPoint and ProxyEndPoint both implement this, so we can easily get the connection security metadata.  SecureRequestCustomerizer uses this connection data and ForwardedRequestCustomizer recreates it to give the request secure status and metadata.

gregw commented:
@joakime I have also updated the heuristic to determine the key length from the cipher name, but we might need to update for any recent ciphers that do not have the key length explicitly in the name like "_256_"

gregw committed:
Improve handling of SSL metadata in ConnectionMetaData, proxied connections and forwarded requests

+ Added an Attributes.Synthetic mechanism to better (and mor commonly) handle additional attributes, whilst meeting the new TCK requirement that they can be deleted or replaced.
+ Made SslSessionData a reusable type provided by EndPoint
+ Added the EndPoint.Securable subtype that can provide a SslSessionData instance for secure EndPoints
+ Updated ForwardedRequestCustomizer, SecureRequestCustomizer to understand Securable and SslSessionData

gregw committed:
Improve handling of SSL metadata in ConnectionMetaData, proxied connections and forwarded requests

+ Added an Attributes.Synthetic mechanism to better (and mor commonly) handle additional attributes, whilst meeting the new TCK requirement that they can be deleted or replaced.
+ Made SslSessionData a reusable type provided by EndPoint
+ Added the EndPoint.Securable subtype that can provide a SslSessionData instance for secure EndPoints
+ Updated ForwardedRequestCustomizer, SecureRequestCustomizer to understand Securable and SslSessionData

gregw committed:
Improve handling of SSL metadata in ConnectionMetaData, proxied connections and forwarded requests

+ Added an Attributes.Synthetic mechanism to better (and mor commonly) handle additional attributes, whilst meeting the new TCK requirement that they can be deleted or replaced.
+ Made SslSessionData a reusable type provided by EndPoint
+ Added the EndPoint.Securable subtype that can provide a SslSessionData instance for secure EndPoints
+ Updated ForwardedRequestCustomizer, SecureRequestCustomizer to understand Securable and SslSessionData

gregw committed:
Improve handling of SSL metadata in ConnectionMetaData, proxied connections and forwarded requests

+ Added an Attributes.Synthetic mechanism to better (and mor commonly) handle additional attributes, whilst meeting the new TCK requirement that they can be deleted or replaced.
+ Made SslSessionData a reusable type provided by EndPoint
+ Added the EndPoint.Securable subtype that can provide a SslSessionData instance for secure EndPoints
+ Updated ForwardedRequestCustomizer, SecureRequestCustomizer to understand Securable and SslSessionData

gregw committed:
Improve handling of SSL metadata in ConnectionMetaData, proxied connections and forwarded requests

+ Added an Attributes.Synthetic mechanism to better (and mor commonly) handle additional attributes, whilst meeting the new TCK requirement that they can be deleted or replaced.
+ Made SslSessionData a reusable type provided by EndPoint
+ Added the EndPoint.Securable subtype that can provide a SslSessionData instance for secure EndPoints
+ Updated ForwardedRequestCustomizer, SecureRequestCustomizer to understand Securable and SslSessionData

gregw committed:
Improve handling of SSL metadata in ConnectionMetaData, proxied connections and forwarded requests

+ Added an Attributes.Synthetic mechanism to better (and mor commonly) handle additional attributes, whilst meeting the new TCK requirement that they can be deleted or replaced.
+ Made SslSessionData a reusable type provided by EndPoint
+ Added the EndPoint.Securable subtype that can provide a SslSessionData instance for secure EndPoints
+ Updated ForwardedRequestCustomizer, SecureRequestCustomizer to understand Securable and SslSessionData

gregw committed:
Improve handling of SSL metadata in ConnectionMetaData, proxied connections and forwarded requests

+ Added an Attributes.Synthetic mechanism to better (and mor commonly) handle additional attributes, whilst meeting the new TCK requirement that they can be deleted or replaced.
+ Made SslSessionData a reusable type provided by EndPoint
+ Added the EndPoint.Securable subtype that can provide a SslSessionData instance for secure EndPoints
+ Updated ForwardedRequestCustomizer, SecureRequestCustomizer to understand Securable and SslSessionData

gregw committed:
Improve handling of SSL metadata in ConnectionMetaData, proxied connections and forwarded requests

+ Added an Attributes.Synthetic mechanism to better (and mor commonly) handle additional attributes, whilst meeting the new TCK requirement that they can be deleted or replaced.
+ Made SslSessionData a reusable type provided by EndPoint
+ Added the EndPoint.Securable subtype that can provide a SslSessionData instance for secure EndPoints
+ Updated ForwardedRequestCustomizer, SecureRequestCustomizer to understand Securable and SslSessionData

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/EndPoint.java:
```diff
@@ -324,5 +326,52 @@ default void write(Callback callback, ByteBuffer... buffers) throws WritePending
+    /**
+     * An endpoint that may be Secure.
+     */
+    interface Securable extends EndPoint
```
Given that `ProxyEndPoint` may or may not be secure, seems like every `EndPoint` could be `Securable` i.e. return `null` from `getSslSessionData()`.
So maybe we just move the methods to `EndPoint`?

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/ProxyConnectionFactory.java:
```diff
@@ -773,6 +802,25 @@ public ProxyEndPoint(EndPoint endPoint, SocketAddress local, SocketAddress remot
+            _sslSessionData = new SslSessionData(null, id, cipher, null);
+        }
+
+        public String getTlsVersion()
```
I would prefer this to be renamed to `getTLSVersion()`.

gregw replied:
OK.  But is it actually used?  I only see accessed in tests.   Can we remove it?

sbordet replied:
You added it in this PR; before it was an attribute used in `SecureRequestCustomizer` to detect whether it was a secure transport.

gregw replied:
I mean do we need the TLS version in some form or other.  If we are just using it to set isSecure, then we don't need to expose the version one way or the other.  I'll remove the method as we now have isSecure

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/EndPoint.java:
```diff
@@ -324,5 +326,52 @@ default void write(Callback callback, ByteBuffer... buffers) throws WritePending
+     * @param peerCertificates The {@link SSLSession#getPeerCertificates()}s converted to {@link X509Certificate}, if known
+     *                         else {@code null}.
+     */
+    record SslSessionData(SSLSession sslSession, String sessionId, String cipherSuite, X509Certificate[] peerCertificates)
```
Rename to `TLSSessionData`, as `SSL` is a thing of the past.
Not sure about being a `record` as it cannot be subclassed (in case we want to add other fields in the future, or allow users to extend).
Should be a class and hide all the machinery (e.g. deriving the `sessionId` string, etc.)?

gregw replied:
I agree about the `record`.  I was going to change to a class, but thought this PR was already big enough.... but I guess that horse had already bolted :(

However, I don't agree about the name.  I know TLS is the current name and SSL is effectively dead.   But everything else is named SSL: the JVM classes, our implementations etc. etc.   So I find it confusing whenever we use TLS.    I think we should stick with SSL until such time as we can do a complete change to TLS everywhere.

sbordet replied:
>  I think we should stick with SSL until such time as we can do a complete change to TLS everywhere.

Then it will never happen, as the JDK classes won't be renamed.

gregw replied:
So we should keep it aligned with our SslConnection* and SslContext* classes.  We might one day rename them.

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/EndPoint.java:
```diff
@@ -324,5 +326,52 @@ default void write(Callback callback, ByteBuffer... buffers) throws WritePending
+         * The name at which an {@code SslSessionData} instance may be found as a request
+         * {@link org.eclipse.jetty.util.Attributes Attribute} or from {@link SSLSession#getValue(String)}.
+         */
+        public static final String ATTRIBUTE = "org.eclipse.jetty.io.Endpoint.SslSessionData";
```
Better `ATTRIBUTE = SslSessionData.class.getName()`.

gregw replied:
Yu'd think!   But if I do that, then the JVM won't let me use the field as a case in a switch statement!!!!

gregw replied:
The string needs to be compile time constant, which the compiler does not know a class name is a constant.    I could put something like the following in `AbstractEndPoint`:
```java
    static
    {
        if (!SslSessionData.class.getName().equals(EndPoint.SslSessionData.ATTRIBUTE))
            throw new IllegalStateException();
    }
```
thoughts?

sbordet replied:
Ah, right about the `switch`. Leave it as a literal.

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/EndPoint.java:
```diff
@@ -324,5 +326,52 @@ default void write(Callback callback, ByteBuffer... buffers) throws WritePending
+         */
+        public static final String ATTRIBUTE = "org.eclipse.jetty.io.Endpoint.SslSessionData";
+
+        public static SslSessionData from(SslSessionData baseData, SSLSession sslSession, String sessionId, String cipherSuite, X509Certificate[] peerCertificates)
```
Never really use all the parameters...
Perhaps use a `with()` pattern such as `newData = baseData.withCipher(cipher)`, or a custom constructor `SslSessionData(SslSessionData base, String cipher)` reads better than the code below.

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ssl/SslConnection.java:
```diff
@@ -168,20 +171,26 @@ public String toString()
+
+    public SslConnection(ByteBufferPool byteBufferPool, Executor executor, EndPoint endPoint, SSLEngine sslEngine, SslContextFactory sslContextFactory)
+    {
+        this(byteBufferPool, executor, endPoint, sslEngine, sslContextFactory, false, false);
```
Move the factory as 3rd param, as it has a longer lifecycle than EndPoint and SSLEngine.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/SecureRequestCustomizer.java:
```diff
@@ -414,29 +316,16 @@ public Object getAttribute(String name)
-     */
-    public record SslSessionData(String sessionId, String cipherSuite, int keySize, X509Certificate[] peerCertificates)
-    {
-    }
```
I fear we're going to break users by removing this class.

gregw replied:
We are not removing it, we are just moving it.     Should have been done in 12.0.0, so it is a little late and a little breaking.  But it will only affect users that have adopted the core API since 12.0.0

However, now that `EndPoint.SslSessionData` is an interface, we could add a deprecated version back.... but it has only been exposed as a public record since 12.  I'd prefer to see if anybody complains first.

sbordet replied:
We're breaking other things, so ok to remove.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/SecureRequestCustomizer.java:
```diff
@@ -29,63 +28,31 @@
-     */
-    public static final String DEFAULT_SSL_SESSION_DATA_ATTRIBUTE = newSslSessionDataAttribute(DEFAULT_SSL_SESSION_ATTRIBUTE);
+    public static final String SSL_SESSION_ATTRIBUTE = "org.eclipse.jetty.server.SslSession";
+    public static final String SSL_SESSION_DATA_ATTRIBUTE = EndPoint.SslSessionData.ATTRIBUTE;
```
The removed constants were available as attribute in the core `Request`, so removing them is a breaking change.

gregw replied:
We only break since 12.0.0, since before that they had "javax" or "jakarta" in their names.    Every attribute has a cost and it is far better to put related values behind a single attribute rather than repeat the mistake of the servlet spec to have multiple individual attributes.

I say we break and if anybody complains we can add a "legacy" handler to rewrap the request and add them back just for that user.

gregw replied:
I have added them back in as an optional legacy support, marked for removal.

sbordet replied:
There is no point deprecating half and removing the other half.
If you're fine with non-backwards compatible changes for 12.0.x, remove what is not used anymore, rather than deprecating.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/ForwardedRequestCustomizer.java:
```diff
@@ -609,6 +610,9 @@ public String toString()
             ? request.getHeaders()
             : HttpFields.build(request.getHeaders(), authority);
 
+        if (forwarded._sslSessionData != null)
+            request.setAttribute(EndPoint.SslSessionData.ATTRIBUTE, forwarded._sslSessionData);
```
Not sure this is right... should not this be done in the wrapper? I don't want to see this attribute if I unwrap the request.

gregw replied:
I have flipped back and forth on this one.   If this was a handler, I'd agree 100%.  But this is a customizer, which is part of preparing the "original" request, which really should never be seen/handled unwrapped.
Also, leaving it in place will make logging data from it simpler in the requestLogger (as currently for things like this we often end up adding an attribute so the request logger can see it).

Let's defer this one until other issues are resolved and we can discuss.

@lachlan-roberts can you look at this PR in regards to request logging?  I can't see anything logging related, but it makes me think there should be???

sbordet replied:
@gregw all the other changes to URI, headers, etc. are wrapped, so it would be good to have attributes wrapped too, but it's far more complicated than just wrapping. I can live with setting the attribute.

gregw replied:
It is not that much more complicated and we are wrapping the request anyway.  It is more what we should do.   I looked at other customizers and they wrap, so unless there is a logging reason not to wrap then I'll wrap.

sbordet commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ssl/SslConnection.java:
```diff
@@ -168,20 +171,26 @@ public String toString()
 
     public SslConnection(ByteBufferPool byteBufferPool, Executor executor, EndPoint endPoint, SSLEngine sslEngine,
-                         boolean useDirectBuffersForEncryption, boolean useDirectBuffersForDecryption)
+                         SslContextFactory sslContextFactory, boolean useDirectBuffersForEncryption, boolean useDirectBuffersForDecryption)
```
This is a breaking change. Can you add another constructor?

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/ForwardedRequestCustomizer.java:
```diff
@@ -872,7 +877,7 @@ private MutableHostPort getFor()
         public void handleCipherSuite(HttpField field)
         {
-            _request.setAttribute("jakarta.servlet.request.cipher_suite", field.getValue());
+            _sslSessionData = EndPoint.SslSessionData.from(_sslSessionData, null, null, field.getValue(), null);
```
Not sure about this. Users will pull out a `SslSessionData`, but unlike the class name says, it will have a `null` `SSLSession`. Not sure we should use the same classes for forward and secure customizers.

gregw replied:
As this PR moved the class (record?) to `EndPoint` we can't have separate classes for how the information was provided.  Besides, the whole point is that the user of this should not know if the data comes from a connection or a proxy or the proxy protocol.

I don't see a problem with `SslSessionData` having a null reference to `SslSession`.  Users understand the difference between metadata and data, and are OK with having one but not the other.    If we had the SslSession, we'd mostly not need the SslSessionData  (other than as the optimization that it was originally written for).

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/ProxyConnectionFactory.java:
```diff
@@ -380,6 +378,23 @@ private enum Transport
+        private static final int PP2_CLIENT_SSL = 0x01;
+        private static final int PP2_CLIENT_CERT_CONN = 0x02;
+        private static final int PP2_CLIENT_CERT_SESS = 0x04;
+
```
Half of these are not used and only clutter the source.
I'd restore what was before.

gregw replied:
There was a bug in the previous code of using `PP2_TYPE_ALPN` rather than `PP2_CLIENT_SSL`, I think because all the values from the spec were not listed.
Having the constants exactly as they are from the spec has value and makes maintenance and understanding simpler.     It is feature that we can see which values are not implemented!

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/ProxyConnectionFactory.java:
```diff
@@ -643,32 +665,42 @@ private void parseBodyAndUpgrade() throws IOException
+                    // TODO should client certificate information be passed? For now we just check there is a client value
+                    //      and the certificate has been verified
+                    if (client != 0 && sslVerified)
+                        proxyEndPoint.setSecure(sslVersion, uniqueId, sslDistinguishedName, sslCipher);
```
I think this additional handling of client certificate data is out of scope for this PR, also because it's sprinkled with TODOs.
Maybe file a new issue, and if it ever get asked we'll implement it.

gregw replied:
That is mostly what I've done, hence the TODO.

The changes I have made here have been to fix bugs.  Specifically we need to check the validate value, else we are saying a connection is secure when it has not been validated.

OK, I also added the handling of cipher name, because that is what this PR is about... common handling of SslSessionData.   The proxy protocol provides it, so we should use it.      Anything we don't have handling for I have TODO'd

sbordet replied:
There were no bugs, you may have introduced them in a previous commit in this PR and then fixed.

The validation is about the client certificate, not about whether the connection is secure, so I don't think we should care about the `verify` field.
Furthermore, it may be that the sender was not able to verify the client certificate, but the receiver is.

The `uniqueId` is not the `SSLSession` id (it has nothing to do with SSL), but it is wrongly used so.

The `distinguishedName` is that of the client certificate, not the SNI.

The only things that can be used to create an `SslSessionData` are the TLS version (which is not reported by `SslSessionData`) and the TLS cipher.

I think the code should be modified to only parse the version and cipher, and `proxyEndPoint.setSecure()` take only the version and cipher, and made non-`public` so that we do not expose it to users in case we change it later.

sbordet commented on jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/ProxyProtocolTest.java:
```diff
@@ -143,14 +143,16 @@ public boolean handle(Request request, Response response, Callback callback)
-                       Arrays.equals(customE1, proxyEndPoint.getTLV(0xE1)) &&
-                       proxyEndPoint.getTLV(0xE2) == null;
+                if (request.getConnectionMetaData().getConnection() instanceof HttpConnection con)
+                {                EndPoint endPoint = con.getEndPoint();
```
Fix formatting.

sbordet commented on jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/ssl/ServerConnectorSslServerTest.java:
```diff
@@ -221,16 +223,15 @@ public boolean handle(Request request, Response response, Callback callback) thr
             out.append("X509Certificate='").append(data == null ? "" : data.peerCertificates()).append("'").append('\n');
             out.append("cipher_suite='").append(session == null ? "" : session.getCipherSuite()).append("'").append('\n');
-            out.append("key_size='").append(data == null ? "" : data.keySize()).append("'").append('\n');
+            out.append("key_size='").append(data == null ? "" : SslContextFactory.deduceKeyLength(data.cipherSuite())).append("'").append('\n');
```
I think data.keySize() should be restored.

gregw replied:
As a field or just as a method?    
I'll add it as a method for now and we will only calculate if it is requested.

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/Attributes.java:
```diff
@@ -135,6 +136,34 @@ static Attributes unwrap(Attributes attributes)
+        if (o instanceof Attributes a)
+        {
+            Set<String> ours = attributes.getAttributeNameSet();
+            Set<String> theirs = attributes.getAttributeNameSet();
```
Should be `a.getAttributeNameSet()`.

The method should be private, not public, or better yet just the default implementation.
 All subinterfaces could just inherit it, since it is based on method that subinterfaces will override.

gregw replied:
fixed bug, but ditto for the default implementation.

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/Attributes.java:
```diff
@@ -135,6 +136,34 @@ static Attributes unwrap(Attributes attributes)
         return attributes;
     }
 
+    static int hashCode(Attributes attributes)
+    {
+        int hash = 113;
+        for (String name : attributes.getAttributeNameSet())
+            hash += name.hashCode() ^ attributes.getAttribute(name).hashCode();
+        return hash;
+    }
```
Just make it the default implementation, less code around, not public, etc.
All subinterfaces could just inherit it, since it is based on method that subinterfaces will override.

gregw replied:
yeah.... but no.  Already tried that and java doesn't let you have a default method that overrides a method in `Object`.

sbordet replied:
Ok, a surprise you can't override.

Then please override `hashCode()` and `equals()` and make them `abstract` (it can be done), so it's clear that implementation must call these `static` methods.

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/Attributes.java:
```diff
@@ -536,40 +561,168 @@ public Set<String> getAttributeNameSet()
-                    return false;
+                // We will need a layer to record modifications to a synthetic attribute
+                if (_layer == null)
+                    _layer = new HashMap<>();
```
Should be thread-safe like `Lazy` is.

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/ssl/SslContextFactory.java:
```diff
@@ -133,6 +136,7 @@ public abstract class SslContextFactory extends ContainerLifeCycle implements Du
         "^.*_NULL_.*$",
         "^.*_anon_.*$"
     };
+    public static final String X_509 = "X.509";
```
Make it `private`.

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/ssl/SslContextFactory.java:
```diff
@@ -1934,36 +1928,27 @@ public static X509Certificate[] getCertChain(SSLSession sslSession)
     }
 
-    private static X509Certificate[] getX509CertChain(SslContextFactory sslContextFactory, SSLSession sslSession)
+    public static X509Certificate[] getX509CertChain(CertificateFactory certificateFactory, SSLSession sslSession)
```
Make it package-protected or restore the original parameters, as passing a `CertificateFactory` is typically complicated for users.

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/ssl/SslContextFactory.java:
```diff
@@ -113,6 +114,8 @@ public abstract class SslContextFactory extends ContainerLifeCycle implements Du
 
     private static final Logger LOG = LoggerFactory.getLogger(SslContextFactory.class);
     private static final Logger LOG_CONFIG = LoggerFactory.getLogger(LOG.getName() + ".config");
+    private static final Pattern DASH_DIGITS_DASH = Pattern.compile("_(\\d+)_");
```
It's `UNDERSCORE_DIGITS_UNDERSCORE`, but rename to `KEY_SIZE_PATTERN`.

sbordet commented on jetty-core/jetty-util/src/test/java/org/eclipse/jetty/util/AttributesTest.java:
```diff
@@ -28,13 +31,40 @@
 public class AttributesTest
 {
+    static Set<String> __syntheticAttributes = Set.of("Roy", "Pris", "Zhora", "Leon");
+    static AtomicReference<Object> __roy = new AtomicReference<>();
+    static AtomicReference<Object> __pris = new AtomicReference<>();
+    static AtomicReference<Object> __zhora = new AtomicReference<>();
+    static AtomicReference<Object> __leon = new AtomicReference<>();
+
     public static Stream<Attributes> attributes()
     {
         return Stream.of(
             new Attributes.Mapped(),
             new Attributes.Wrapper(new Attributes.Mapped()),
             new Attributes.Layer(new Attributes.Mapped()),
-            new AttributesMap()
+            new AttributesMap(),
+            new Attributes.Synthetic(new Attributes.Mapped())
+            {
+                @Override
+                protected Object getSyntheticAttribute(String name)
+                {
+                    return switch (name)
+                    {
+                        case "Roy" -> __roy.get();
+                        case "Pris" -> __pris.get();
+                        case "Zhora" -> __zhora.get();
+                        case "Leon" -> __leon.get();
+                        default -> null;
+                    };
+                }
+
+                @Override
+                protected Set<String> getSyntheticNameSet()
+                {
+                    return __syntheticAttributes;
+                }
+            }
```
Seems overcomplicated.
The existing test will not test any `Synthetic` feature, and `Wrapper` is already tested, and the new test will skip all the others.
I would just make `testSynthetic()` a non-parametric test and do all the testing locally there, removing all unnecessary statics (that also are not good for parallel testing).

gregw replied:
I want to test that a `Synthetic` passes all the other behaviours exactly as any other.  This unit test is written as it is exactly because we have introduced different behaviours before.
I'll move the statics to be contained in a TestSynthentAttributes classs.....

sbordet commented on jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletContextRequest.java:
```diff
@@ -58,15 +58,27 @@
+    public static final String SSL_CIPHER_SUITE = "jakarta.servlet.request.cipher_suite";
+    public static final String SSL_KEY_SIZE = "jakarta.servlet.request.key_size";
+    public static final String SSL_SESSION_ID = "jakarta.servlet.request.ssl_session_id";
+    public static final String X_509_CERTIFICATE = "jakarta.servlet.request.X509Certificate";
```
This attribute is mis-named by the spec, since it returns the certificates (plural).
I'd rename to `X509_CERTIFICATES`.

sbordet commented on jetty-ee10/jetty-ee10-servlet/src/test/java/org/eclipse/jetty/ee10/servlet/RequestTest.java:
```diff
@@ -334,62 +336,50 @@ protected void service(HttpServletRequest request, HttpServletResponse resp) thr
-                    "jakarta.servlet.request.ssl_session_id",
-                    SecureRequestCustomizer.PEER_CERTIFICATES_ATTRIBUTE,
-                    "jakarta.servlet.request.X509Certificate",
+                    SecureRequestCustomizer.SSL_SESSION_DATA_ATTRIBUTE,
+                    ServletContextRequest.SSL_CIPHER_SUITE,
+                    ServletContextRequest.SSL_KEY_SIZE,
+                    ServletContextRequest.SSL_SESSION_ID,
+                    ServletContextRequest.X_509_CERTIFICATE,
```
As with other tests, I don't think we can break backwards compatibility.

gregw replied:
Are you worried about no longer supporting both `SecureRequestCustomizer.SSL_SESSION_ID_ATTRIBUTE` and `"jakarta.servlet.request.ssl_session_id"`??

We are kind of doomed to make a breaking change in 12 since from in 10 and 11 the `SecureRequestCustomizer` was setting the servlet specific name with "javax" or "jakarta".   So it was only since 12.0.0 we have had a neutral version.   I don't think it is so bad to replace those individual ones with a single attribute now..... If somebody complains we can put them back.  If I put them back now, they are there forever, even if deprecated.

sbordet commented on jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/Request.java:
```diff
@@ -105,6 +105,12 @@
 public class Request implements HttpServletRequest
 {
     public static final String __MULTIPART_CONFIG_ELEMENT = "org.eclipse.jetty.multipartConfig";
+    public static final String JAKARTA_SERVLET_REQUEST_PREFIX = "jakarta.servlet.request.";
+    public static final String MULTIPART_CONFIG_ELEMENT = "org.eclipse.jetty.multipartConfig";
+    public static final String SSL_CIPHER_SUITE = "jakarta.servlet.request.cipher_suite";
+    public static final String SSL_KEY_SIZE = "jakarta.servlet.request.key_size";
+    public static final String SSL_SESSION_ID = "jakarta.servlet.request.ssl_session_id";
+    public static final String X_509_CERTIFICATE = "jakarta.servlet.request.X509Certificate";
```
Breaks backwards compatibility.

gregw replied:
Which bit?  I've just added constants... Oh I see the old `__` version.  I'll deprecate that one as it should never of had the `__` prefix if public.

sbordet reviewed (changes requested):
Almost impossible to review.
Next time, please split it.

gregw committed:
Merge branch 'jetty-12.0.x' into fix/jetty-12.0.x/10775/ConnectionMetaDataIsSecure

gregw committed:
Improve handling of SSL metadata in ConnectionMetaData, proxied connections and forwarded requests

+ Added an Attributes.Synthetic mechanism to better (and mor commonly) handle additional attributes, whilst meeting the new TCK requirement that they can be deleted or replaced.
+ Made SslSessionData a reusable type provided by EndPoint
+ Added the EndPoint.Securable subtype that can provide a SslSessionData instance for secure EndPoints
+ Updated ForwardedRequestCustomizer, SecureRequestCustomizer to understand Securable and SslSessionData

gregw committed:
Improve handling of SSL metadata in ConnectionMetaData, proxied connections and forwarded requests

+ Added an Attributes.Synthetic mechanism to better (and mor commonly) handle additional attributes, whilst meeting the new TCK requirement that they can be deleted or replaced.
+ Made SslSessionData a reusable type provided by EndPoint
+ Added the EndPoint.Securable subtype that can provide a SslSessionData instance for secure EndPoints
+ Updated ForwardedRequestCustomizer, SecureRequestCustomizer to understand Securable and SslSessionData

gregw commented:
@sbordet yeah apologies again for the monster PR.  kind of got away from me.

I know there are some breaking API changes in this, but they only break things that were added (badly) in 12.0.0.    We repeated some mistakes of the servlet spec (too many individual attributes rather than one bundle of related attributes) and it was also to specific the source of the data (the user does not care if isSecure or session ID comes from connection, proxy request or proxy protocol).

I'm happy to live with these breaking changes.  If anybody complains, we can easily write a handler to wrap the request again with these 12-legacy attributes.

gregw committed:
Various cleanups of Handler.insertHandler

gregw committed:
Merge branch 'jetty-12.0.x' into fix/jetty-12.0.x/10775/ConnectionMetaDataIsSecure

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/ProxyConnectionFactory.java:
```diff
@@ -773,6 +802,25 @@ public ProxyEndPoint(EndPoint endPoint, SocketAddress local, SocketAddress remot
+            return _sslSessionData;
+        }
+
+        public void setSecure(String version, String id, String distinguishedName, String cipher)
```
I'd prefer to have this method be non-public and be `setSslSessionData()`, with the caller creating the `SslSessionData`.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/SecureRequestCustomizer.java:
```diff
@@ -29,68 +27,59 @@
     private long _stsMaxAge;
     private boolean _stsIncludeSubDomains;
     private HttpField _stsField;
+    private final boolean _legacyAttributes;
```
Remove this.

sbordet commented on jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/ProxyProtocolTest.java:
```diff
@@ -143,14 +143,17 @@ public boolean handle(Request request, Response response, Callback callback)
-                return Arrays.equals(customE0, proxyEndPoint.getTLV(0xE0)) &&
-                       Arrays.equals(customE1, proxyEndPoint.getTLV(0xE1)) &&
-                       proxyEndPoint.getTLV(0xE2) == null;
+                if (request.getConnectionMetaData().getConnection() instanceof HttpConnection con)
```
No need for this `if`, just use `EndPoint endPoint = request.getConnectionMetaData().getConnection().getEndPoint();`.

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/Attributes.java:
```diff
@@ -135,6 +136,34 @@ static Attributes unwrap(Attributes attributes)
         return attributes;
     }
 
+    static int hashCode(Attributes attributes)
+    {
+        int hash = 113;
+        for (String name : attributes.getAttributeNameSet())
+            hash += name.hashCode() ^ attributes.getAttribute(name).hashCode();
+        return hash;
+    }
```
I don't think we want to be fancy and deviate from widely-known hash functions.

```suggestion
    static int hashCode(Attributes attributes)
    {
        int hash = 0;
        for (String name : attributes.getAttributeNameSet())
        {
            hash = 31 * hash + Objects.hash(name, attributes.get(name));
        }
        return hash;
    }
```
FTR, `31` seems not a good number either (https://stackoverflow.com/questions/1835976/what-is-a-sensible-prime-for-hashcode-calculation/2816747#2816747), but it's what Java uses and the algorithm is right.

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/Attributes.java:
```diff
@@ -203,27 +232,25 @@ public void clearAttributes()
+        public boolean equals(Object o)
         {
-            return getWrapped().equals(obj);
+            return o instanceof Attributes && Attributes.equals(this, o);
```
```suggestion
            return Attributes.equals(this, o);
```

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/Attributes.java:
```diff
@@ -305,28 +332,16 @@ public Set<Map.Entry<String, Object>> getAttributeEntrySet()
-                return true;
-            }
-            return false;
+            return o instanceof Attributes && Attributes.equals(this, o);
```
```suggestion
            return Attributes.equals(this, o);
```

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/Attributes.java:
```diff
@@ -442,14 +457,25 @@ public void dump(Appendable out, String indent) throws IOException
+        @Override
+        public boolean equals(Object o)
+        {
+            return o instanceof Attributes && Attributes.equals(this, o);
```
```suggestion
            return Attributes.equals(this, o);
```

gregw replied:
tell intellij!

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/Attributes.java:
```diff
@@ -609,5 +761,17 @@ public Map<String, Object> asAttributeMap()
             return Collections.emptyMap();
         }
+
+        @Override
+        public int hashCode()
+        {
+            return Attributes.hashCode(this);
+        }
+
+        @Override
+        public boolean equals(Object o)
+        {
+            return o instanceof Attributes && Attributes.equals(this, o);
+        }
```
These are not necessary, given how `Wrapper` has been implemented (by not delegating).

gregw replied:
tell intellij that

sbordet commented on jetty-core/jetty-util/src/test/java/org/eclipse/jetty/util/AttributesTest.java:
```diff
@@ -135,4 +139,88 @@ public void testAttributeLayer()
         testAttributes(layer);
     }
 
+    @ParameterizedTest
+    @MethodSource("attributes")
+    public void testSynthetic(Attributes attributes)
+    {
+        Assumptions.assumeTrue(attributes instanceof Attributes.Synthetic);
```
Just get rid of the parametrization, allocate `TestSynthetic` and test only that.

sbordet committed:
Avoid using deprecated SslConnection constructors.
Changed signature of SslClientConnectionFactory.newSslConnection() to reflect the need to pass the SslContextFactory (and added getters for required components).
Renamed SslSessionData.sessionId() to sslSessionId() for clarity.
Renamed Request.X_509_CERTIFICATES to PEER_CERTIFICATES for clarity.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

gregw committed:
Improve handling of SSL metadata in ConnectionMetaData, proxied connections and forwarded requests

+ Added an Attributes.Synthetic mechanism to better (and mor commonly) handle additional attributes, whilst meeting the new TCK requirement that they can be deleted or replaced.
+ Made SslSessionData a reusable type provided by EndPoint
+ Added the EndPoint.Securable subtype that can provide a SslSessionData instance for secure EndPoints
+ Updated ForwardedRequestCustomizer, SecureRequestCustomizer to understand Securable and SslSessionData

gregw committed:
Improve handling of SSL metadata in ConnectionMetaData, proxied connections and forwarded requests

+ Added an Attributes.Synthetic mechanism to better (and mor commonly) handle additional attributes, whilst meeting the new TCK requirement that they can be deleted or replaced.
+ Made SslSessionData a reusable type provided by EndPoint
+ Added the EndPoint.Securable subtype that can provide a SslSessionData instance for secure EndPoints
+ Updated ForwardedRequestCustomizer, SecureRequestCustomizer to understand Securable and SslSessionData

gregw committed:
Improve handling of SSL metadata in ConnectionMetaData, proxied connections and forwarded requests

+ Added an Attributes.Synthetic mechanism to better (and mor commonly) handle additional attributes, whilst meeting the new TCK requirement that they can be deleted or replaced.
+ Made SslSessionData a reusable type provided by EndPoint
+ Added the EndPoint.Securable subtype that can provide a SslSessionData instance for secure EndPoints
+ Updated ForwardedRequestCustomizer, SecureRequestCustomizer to understand Securable and SslSessionData

sbordet committed:
Javadocs improvements.
Removed unnecessary attribute constants in SecureRequestCustomizer.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>
